File: samples/sys/idle.py
Callouts = []

class Callout:
  def __init__(self, function, time, val = 0):
    self.once = val
    self.func = function
    self.next_call = now + time
    self.next_time = time

  def once_only(self):
    if self.once:
      return 1
    else:
      return 0
  
  def call(self):
    self.func()
    self.next_call = now + self.next_time

  def ready(self):
    if self.next_call <= now:
      return 1
    else:
      return 0

def idle():
  for i in range(len(Callouts)):
    if Callouts[i].ready():
      Callouts[i].call()
      if Callouts[i].once_only():
        del Callouts[i]

def callout_add(function, time):
  callout = Callout(function, time)
  Callouts.append(callout)

def callout_remove(function):
  for i in range(len(Callouts)):
    if Callouts[i].func is function:
      del Callouts[i]

def callout_once(function, time):
  callout = Callout(function, time, 1)
  Callouts.append(callout)

def cmd_show_callouts():
  for i in range(len(Callouts)):
    print "%-20s %6d %6d\n" % (Callouts[i].func, Callouts[i].next_time, Callouts[i].next_call) 

File: samples/sys/idle.pl
# Callout management

@Callouts = ();
@CalloutTimes = ();
@NextCallout = ();

sub idle {
    my ($i);
    for ($i = 0; $i <= $#Callouts; $i++) {
        if ($NextCallout[$i] <= $now) {
            $NextCallout[$i] = $now + $CalloutTimes[$i];
            &{$Callouts[$i]};
        }
    }
}

sub callout_add ($$) {
    my ($function, $time) = @_;
    push @Callouts, $function;
    push @CalloutTimes, $time;
    push @NextCallout, $now + $time;
}

sub callout_once ($$) {
    my ($function, $time, $sub) = @_;
    $sub = sub { callout_remove($sub); &$function(); };
    callout_add ($sub, $time);
}

sub callout_remove ($) {
    my ($what, $x) = @_;
    for ($x = 0; $x <= $#Callouts; $x++) {
        # Is this the right way to compare references?
        if ($Callouts[$x] eq $what) {
            splice(@Callouts, $x, 1);
            splice(@CalloutTimes, $x, 1);
            splice(@NextCallout, $x, 1);
            return;
        }
    }

    print "callout_remove(): Removal of $what FAILED!\n";
}

sub cmd_show_callouts {
    my ($i);
    foreach $i (0 .. $#Callouts) {
        printf "%-20s %6d %6d\n", $Callouts[$i], $CalloutTimes[$i], $NextCallout[$i];
    }
}

1;
